### ISSUE TITLE
iOS/Apple, Surface Pro - not able to receive the stream (the root cause provided)

### ISSUE THREAD

**Issue by greko6** [pion/ion-sfu#23]
### Your environment.
  - Browser: Safari 13.1.2 with Chrome on Windows 
  - Other Information - see below

### What did you do?
1. Opened Chrome on Windows and started the stream.
2. Opened Safari or Chrome on macOS and couldn't receive the stream.

### What did you expect?
See the stream.

### What happened?
I couldn't see it.

After https://github.com/pion/ion-sdk-js/pull/28 fix, I was able to experiment a bit more with Safari.

It turns out, that in the SDP Offer Safari has:
![image](https://user-images.githubusercontent.com/1009724/87076292-68955280-c1d6-11ea-818d-4197441b3069.png)

(...)
m=video 9 UDP/TLS/RTP/SAVPF 96 97 98 99 100 101 127 125 104
(...)
a=rtpmap:96 H264/90000
(...)

but in the SDP answer we have:
m=video 9 UDP/TLS/RTP/SAVPF 96
(...)
a=rtpmap:96 VP8/90000
--------

For Safari, to get VP8/90000, the SAVPF should be set to 100. 

96 works flawlessly with Chrome (on other OS than Mac/iOS), as with Chrome VP8/90000, indeed, it is 96.

To fix that, for maximum compatibility in the future, we probably need to look-up VP8 in the SDP Offer and return the same value to the browser.

Potential solution/fix/help here: https://github.com/pion/webrtc/issues/716#issuecomment-502609027

**Comment by greko6** [comment]
Also, this is the reason for the change: https://groups.google.com/forum/#!topic/discuss-webrtc/Z5YOWMC8TwU

It looks like Safari is using WebRTC before M57. Hence the reason why 100, but not 96.

Key quote from that thread: "External apps need to update their code to not rely on hardcoded values. This is all in accordance with the spec, payload types in the range 96-127 are dynamic and are not tied to any specific codec."

Most likely types will change again in the future.

**Comment by sibis** [comment]
@greko6 this issue exists iOS platform mobile and iPad in general. Since even chrome in iphone and ipad uses WebKit rendering engine not V8 just the chrome UI. So any app built on ion-sdk-js is not working on iphone and ipad irrespective of the browser

**Comment by sibis** [comment]
@tarrencev are we updating the docker image anytime now to include this fix?

**Comment by greko6** [comment]
I did compile it and created a new docker for my own SFU for testing purposes - after the patch was applied.

Unfortunately, SFU is still not matching it. Here is what I can see:

Offer: 
![image](https://user-images.githubusercontent.com/1009724/87887060-761cba80-c9d6-11ea-93a4-745e9a2fedb5.png)

Answer:
![image](https://user-images.githubusercontent.com/1009724/87887079-a6645900-c9d6-11ea-89be-c379ee18334e.png)


The answer should be:

m=video (...)
(...)
a=rtmpmap: 100 VP8/9000

With 96 Safari is still interpreting the video track as h264.

**Comment by greko6** [comment]
Surface Pro tablet is also affected by the same bug. As it's not fixed, can you reopen it, please?

**Comment by greko6** [comment]
@tarrencev could you re-open this issue, please? That way people that are using sfu will have much better visibility it's not working.

**Comment by Sean-Der** [comment]
Hey @greko6 if you go to https://github.com/orgs/pion you should have an invite to join the org!

If you aren't able to join I will re-open, but hard for me to find bandwidth to help with Ion day-to-day :(

**Comment by greko6** [comment]
@Sean-Der : I think one of the GitHub org limitations/features is that other org members cannot re-open issues. Only the person that closed it can. 

Source: https://stackoverflow.com/questions/21333654/how-to-re-open-an-issue-in-github

"You cannot re-open your own issues if a repo collaborator closed them"

**Comment by greko6** [comment]
@Sean-Der : This issue, once solved, will allow pion to drastically accelerate as currently, people that are testing it with MacOSes/iPhones/iPads are not getting the streams at all. More recently, I've learned that also Microsoft Surface Pro is affected because of the same simple mismatch reasons.

**Comment by greko6** [comment]
Update based on the Slack thread.

TL;DR most likely we'll need to wait for v3

From Jeremia: 
"The main issue here is that in the current version of pion/webrtc the PayloadTypes are defined in the webrtc.MediaEngine and are usually either populated from SDP or defined before any tracks are added. Populating from SDP does not scale well if you need to dynamically add tracks after the first negotiation.
The easiest solution for this is to generate the offers on the server-side and then the browsers will use server-defined PayloadTypes in their answers.
I think there are plans for v3 to fix this behavior."

**Comment by tarrencev** [comment]
this is working now but only on `ion-sfu` not with `ion`. you can verify it by running `ion-sfu` with the `pub-sub-in-browser` example. you will need a reverse proxy for ssl though since safari doesn't like ws

**Comment by greko6** [comment]
@tarrencev : Nice! Did you receive a stream on Safari? Which diff fixed it?
